## Data Cleaning Pipeline (Kaggle->Client-Ready)

# Importing Libraries 

In [66]:
import pandas as pd
import numpy as np

# Loading CSV file

In [67]:
df = pd.read_csv("../data/raw_data.csv")

In [68]:
df.shape

(20, 11)

In [69]:
df.head(5)

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,1,1,2,"$780,000,000","$780,000,000",Taylor Swift,The Eras Tour †,2023–2024,56,"$13,928,571",[1]
1,2,1,7[2],"$579,800,000","$579,800,000",Beyoncé,Renaissance World Tour,2023,56,"$10,353,571",[3]
2,3,1[4],2[5],"$411,000,000","$560,622,615",Madonna,Sticky & Sweet Tour ‡[4][a],2008–2009,85,"$4,835,294",[6]
3,4,2[7],10[7],"$397,300,000","$454,751,555",Pink,Beautiful Trauma World Tour,2018–2019,156,"$2,546,795",[7]
4,5,2[4],NaN,"$345,675,146","$402,844,849",Taylor Swift,Reputation Stadium Tour,2018,53,"$6,522,173",[8]


In [70]:
df.tail(5)

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
15,16,NaN,NaN,"$184,000,000","$227,452,347",Pink,The Truth About Love Tour,2013–2014,142,"$1,295,775",[22]
16,17,NaN,NaN,"$170,000,000","$213,568,571",Lady Gaga,Born This Way Ball,2012–2013,98,"$1,734,694",[7]
17,18,NaN,NaN,"$169,800,000","$207,046,755",Madonna,Rebel Heart Tour,2015–2016,82,"$2,070,732",[4]
18,19,NaN,NaN,"$167,700,000[e]","$204,486,106",Adele,Adele Live 2016,2016–2017,121,"$1,385,950",[25]
19,20,NaN,NaN,"$150,000,000","$185,423,109",Taylor Swift,The Red Tour,2013–2014,86,"$1,744,186",[26]


In [71]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype
---  ------                            --------------  -----
 0   Rank                              20 non-null     int64
 1   Peak                              9 non-null      str  
 2   All Time Peak                     6 non-null      str  
 3   Actual gross                      20 non-null     str  
 4   Adjusted gross (in 2022 dollars)  20 non-null     str  
 5   Artist                            20 non-null     str  
 6   Tour title                        20 non-null     str  
 7   Year(s)                           20 non-null     str  
 8   Shows                             20 non-null     int64
 9   Average gross                     20 non-null     str  
 10  Ref.                              20 non-null     str  
dtypes: int64(2), str(9)
memory usage: 1.8 KB


In [72]:
df.describe()

,Rank,Shows
count,20.000000,20.000000
mean,10.450000,110.000000
std,5.942488,66.507617
min,1.000000,41.000000
25%,5.750000,59.000000
50%,10.500000,87.000000
75%,15.250000,134.500000
max,20.000000,325.000000


In [73]:
df.dtypes

Rank                                int64
Peak                                  str
All Time Peak                         str
Actual gross                          str
Adjusted gross (in 2022 dollars)      str
Artist                                str
Tour title                            str
Year(s)                               str
Shows                               int64
Average gross                         str
Ref.                                  str
dtype: object

In [74]:
df.columns

Index(['Rank', 'Peak', 'All Time Peak', 'Actual gross',
       'Adjusted gross (in 2022 dollars)', 'Artist', 'Tour title', 'Year(s)',
       'Shows', 'Average gross', 'Ref.'],
      dtype='str')

## DAMAGE REPORT
This dataset contains information about the top 20 highest-grossing concert tours. It includes rankings, artists, tour names, revenue, number of shows, and years.

## Dataset Shape

- Rows: **20**
- Columns: **11**

## Data Types

The dataset contains two numerical columns and nine string columns. Revenue columns are stored as text due to currency symbols and commas.

## Missing Values

Missing values are present in the **Peak** and **All Time Peak** columns. The remaining columns are complete.

## Statistical Summary

- Average Rank: **10.45**
- Average Shows: **110**
- Maximum Shows: **325**
- Minimum Shows: **41**

## Initial Observations

- Taylor Swift appears multiple times in the dataset.
- Revenue columns need cleaning before analysis.
- The number of shows varies greatly among tours.
- Some ranking information is unavailable for certain tours.

## Working On Duplicate Rows

Identify Duplicate Rows

In [75]:
df.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
dtype: bool

No Duplicate Rows Found hence no dropping of duplicate rows will occur

## Handling Missing Values

In [76]:
df.isnull()

,Rank,Peak,All Time Peak,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False
4,False,False,True,False,False,False,False,False,False,False,False
5,False,False,False,False,False,False,False,False,False,False,False
6,False,False,True,False,False,False,False,False,False,False,False
7,False,True,True,False,False,False,False,False,False,False,False
8,False,True,True,False,False,False,False,False,False,False,False
9,False,True,True,False,False,False,False,False,False,False,False


Missing Values in **Peak** and **All Time Peak**

**Counting Missing Values**

In [77]:
df.isnull().sum()

Rank                                 0
Peak                                11
All Time Peak                       14
Actual gross                         0
Adjusted gross (in 2022 dollars)     0
Artist                               0
Tour title                           0
Year(s)                              0
Shows                                0
Average gross                        0
Ref.                                 0
dtype: int64

### **Peak** has *11* missing values and **All Time Peak** has *14* missing values
### Peak (55%) is missing data and All Time Peak (70%) is a big percentage so its better to drop these columns rather than filling them as it will be called fabricating the data, not cleaning it and if we drop the affected rows it will leave us with little sample data to analyze the dataset further.

## Dropping **"Peak"** and **"All Time Peak"**

In [78]:
df.drop(columns=["Peak", "All Time Peak"],inplace = True)

## Fixing Inconsistent Formats

### Standralizing Test Casing for All String Columns

In [79]:
cols = ["Artist", "Tour title"]

df[cols] = df[cols].apply(lambda col: col.str.strip().str.title())

In [80]:
for col in df.columns:
    print(repr(col))

'Rank'
'Actual\xa0gross'
'Adjusted\xa0gross (in 2022 dollars)'
'Artist'
'Tour title'
'Year(s)'
'Shows'
'Average gross'
'Ref.'


### Fixing DataTypes

In [81]:
c = ["Actual gross", "Adjusted gross (in 2022 dollars)", "Average gross","Ref."]

df[c] = df[c].apply(
    lambda col: pd.to_numeric(
        col.str.replace(r"[^\d]", "", regex=True),
        errors="coerce"
    )
)


## Rechecking DataSet

In [82]:
df.shape

(20, 9)

In [83]:
df.columns

Index(['Rank', 'Actual gross', 'Adjusted gross (in 2022 dollars)', 'Artist',
       'Tour title', 'Year(s)', 'Shows', 'Average gross', 'Ref.'],
      dtype='str')

In [84]:
df.head(5)

,Rank,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
0,1,780000000,780000000,Taylor Swift,The Eras Tour †,2023–2024,56,13928571,1
1,2,579800000,579800000,Beyoncé,Renaissance World Tour,2023,56,10353571,3
2,3,411000000,560622615,Madonna,Sticky & Sweet Tour ‡[4][A],2008–2009,85,4835294,6
3,4,397300000,454751555,Pink,Beautiful Trauma World Tour,2018–2019,156,2546795,7
4,5,345675146,402844849,Taylor Swift,Reputation Stadium Tour,2018,53,6522173,8


In [85]:
df.tail(5)

,Rank,Actual gross,Adjusted gross (in 2022 dollars),Artist,Tour title,Year(s),Shows,Average gross,Ref.
15,16,184000000,227452347,Pink,The Truth About Love Tour,2013–2014,142,1295775,22
16,17,170000000,213568571,Lady Gaga,Born This Way Ball,2012–2013,98,1734694,7
17,18,169800000,207046755,Madonna,Rebel Heart Tour,2015–2016,82,2070732,4
18,19,167700000,204486106,Adele,Adele Live 2016,2016–2017,121,1385950,25
19,20,150000000,185423109,Taylor Swift,The Red Tour,2013–2014,86,1744186,26


In [86]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 9 columns):
 #   Column                            Non-Null Count  Dtype
---  ------                            --------------  -----
 0   Rank                              20 non-null     int64
 1   Actual gross                      20 non-null     int64
 2   Adjusted gross (in 2022 dollars)  20 non-null     int64
 3   Artist                            20 non-null     str  
 4   Tour title                        20 non-null     str  
 5   Year(s)                           20 non-null     str  
 6   Shows                             20 non-null     int64
 7   Average gross                     20 non-null     int64
 8   Ref.                              20 non-null     int64
dtypes: int64(6), str(3)
memory usage: 1.5 KB


In [87]:
df.describe()

,Rank,Actual gross,Adjusted gross (in 2022 dollars),Shows,Average gross,Ref.
count,20.000000,2.000000e+01,2.000000e+01,20.000000,2.000000e+01,20.000000
mean,10.450000,2.879509e+08,3.438781e+08,110.000000,3.726571e+06,87.300000
std,5.942488,1.563284e+08,1.514627e+08,66.507617,3.393340e+06,336.362915
min,1.000000,1.500000e+08,1.854231e+08,41.000000,6.153850e+05,1.000000
25%,5.750000,1.915000e+08,2.457557e+08,59.000000,1.647508e+06,6.750000
50%,10.500000,2.397500e+08,2.974889e+08,87.000000,2.342100e+06,11.500000
75%,15.250000,3.152876e+08,3.924451e+08,134.500000,4.933024e+06,19.250000
max,20.000000,7.800000e+08,7.800000e+08,325.000000,1.392857e+07,1516.000000


### DataSet Cleaned Successfully